src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java

```diff
@@ -288,6 +304,62 @@ public void channelCreated(Channel ch) {
     this.extraHttpHeaders = extraHttpHeaders;
     this.verifyDownloads = verifyDownloads;
     this.digestUtil = digestUtil;
+    this.retrier = retrier != null ? retrier : newRetrier(null);
+  }
+
+  public static RemoteRetrier newRetrier(RemoteOptions options) {
+    CircuitBreaker circuitBreaker = Retrier.ALLOW_ALL_CALLS;
+    if (options != null && options.remoteMaxFailures > 0) {
+      circuitBreaker = new CircuitBreaker() {
+        State state = State.ACCEPT_CALLS;
+        AtomicInteger failures = new AtomicInteger(0);
+        final int maxFailures = options.remoteMaxFailures;
+        @Override
+        public State state() {
+          return state;
+        }
+
+        @Override
+        public void recordFailure() {
+          if (failures.incrementAndGet() > maxFailures) {
+            state = State.REJECT_CALLS;
+          }
+        }
+
+        @Override
+        public void recordSuccess() { }
+      };
+    }
+
+    if (options == null || options.remoteMaxRetryAttempts <= 0) {
+      return new RemoteRetrier(
+          () -> Retrier.RETRIES_DISABLED,
+          (Exception e) -> false,
+          MoreExecutors.listeningDecorator(Executors.newScheduledThreadPool(1)),
+          circuitBreaker);
+    }
+    return new RemoteRetrier(
+        () -> new RemoteRetrier.ExponentialBackoff(options),
+        (Exception e) -> shouldRetry(e),
+        MoreExecutors.listeningDecorator(Executors.newScheduledThreadPool(1)),
+        circuitBreaker);
+  }
+
+  private static boolean shouldRetry(Exception e) {
+    if (e instanceof HttpException) {
+      HttpResponse response = ((HttpException) e).response();
+      return response.status().equals(HttpResponseStatus.INTERNAL_SERVER_ERROR)
+          || response.status().equals(HttpResponseStatus.BAD_GATEWAY);
+    } else if (e instanceof ClosedChannelException
+        || e instanceof ReadTimeoutException
+        || e instanceof WriteTimeoutException) {
+      return true;
+    }
+    else if (e instanceof IOException) {
+      String message = e.getMessage();
+      return RETRYABLE_ERROR_MESSAGE.asPredicate().test(message);
+    }
+    return false;
```

Comment:
What exceptions do we want to leave out? Aside from `HttpResponseStatus.NOT_FOUND` for which we don't want to retry, is there anything else? Wouldn't be easier/better if we return `false` for the exceptions we don't want to retry and retry for any other exception?

Reply:
I'd rather keep it as a whitelist of retryable errors. There are many errors that should not be considered transient, although I'd expect most of them to not be involved in this specific case. Still, since retries are silently executed while error surface as warnings, I'd prefer to update the list if necessary.

Reply:
okidoki